packages/lib/test/computedTree/computedTree.test.ts

```diff
@@ -0,0 +1,518 @@
+
+  expect(() => r.modelIsParent).toThrow("Maximum call stack size exceeded")
+})
+
+test("computed tree node's onAttachedToRootStore hook is called", () => {
+  const r = new R({})
+
+  expect(r.model).toBeInstanceOf(M)
+
+  expect(r.model.isAttachedToRootStore).toBeFalsy()
+
+  registerRootStore(r)
+  expect(r.model.isAttachedToRootStore).toBeTruthy()
+
+  unregisterRootStore(r)
+  expect(r.model.isAttachedToRootStore).toBeFalsy()
+})
+
+describe("tree traversal functions", () => {
+  test("computed tree supports getParentPath", () => {
+    const r = new R({})
+
+    expect(getParentPath(r.array)).toEqual({ parent: r, path: "array" })
+    expect(getParentPath(r.plainObject)).toEqual({ parent: r, path: "plainObject" })
+    expect(getParentPath(r.model)).toEqual({ parent: r, path: "model" })
+  })
+
+  test("computed tree supports getParent", () => {
+    const r = new R({})
+
+    expect(getParent(r.array)).toBe(r)
+    expect(getParent(r.plainObject)).toBe(r)
+    expect(getParent(r.model)).toBe(r)
+  })
+
+  test("computed tree supports getParentToChildPath", () => {
+    const r = new R({})
+
+    expect(getParentToChildPath(r, r.array)).toEqual(["array"])
+    expect(getParentToChildPath(r, r.plainObject)).toEqual(["plainObject"])
+    expect(getParentToChildPath(r, r.model)).toEqual(["model"])
+  })
+
+  test("computed tree supports getRootPath", () => {
+    const r = new R({})
+
+    expect(getRootPath(r.array).path).toEqual(["array"])
+    expect(getRootPath(r.plainObject).path).toEqual(["plainObject"])
+    expect(getRootPath(r.model).path).toEqual(["model"])
+  })
+
+  test("computed tree supports getRoot", () => {
+    const r = new R({})
+
+    expect(getRoot(r.array)).toBe(r)
+    expect(getRoot(r.plainObject)).toBe(r)
+    expect(getRoot(r.model)).toBe(r)
+  })
+
+  test("computed tree supports isRoot", () => {
+    const r = new R({})
+
+    expect(isRoot(r.array)).toBeFalsy()
+    expect(isRoot(r.plainObject)).toBeFalsy()
+    expect(isRoot(r.model)).toBeFalsy()
+  })
+
+  test("computed tree supports isChildOfParent", () => {
+    const r = new R({})
+
+    expect(isChildOfParent(r.array, r)).toBeTruthy()
+    expect(isChildOfParent(r.plainObject, r)).toBeTruthy()
+    expect(isChildOfParent(r.model, r)).toBeTruthy()
+  })
+
+  test("computed tree supports isParentOfChild", () => {
+    const r = new R({})
+
+    expect(isParentOfChild(r, r.array)).toBeTruthy()
+    expect(isParentOfChild(r, r.plainObject)).toBeTruthy()
+    expect(isParentOfChild(r, r.model)).toBeTruthy()
+  })
+
+  // fails because `resolvePath` skips ahead to `<model>.$` when it arrives at a model node
+  test.skip("computed tree supports resolvePath", () => {
+    const r = new R({})
+
+    expect(resolvePath(r, ["array"])).toEqual({ resolved: true, value: r.array })
+    expect(resolvePath(r, ["plainObject"])).toEqual({ resolved: true, value: r.plainObject })
+    expect(resolvePath(r, ["model"])).toEqual({ resolved: true, value: r.model })
```

Comment:
perhaps resolvePath should be changed to also check the original object (not the $ property) but only if nothing is found inside the $ property before "failing"?

Reply:
I've changed `resolvePath` accordingly (d09dc8b406568e888ffdeccd1856d7bf8cbe8f2c).